File: backend/services/converter.py
import pandas as pd
from typing import Dict, Any
from cryptography.fernet import Fernet
from datetime import datetime
import xml.etree.ElementTree as ET
from xml.dom import minidom
import os
from dotenv import load_dotenv
import logging
import re

# Load environment variables
load_dotenv()

# Initialize logger
log = logging.getLogger("converter")

class ExcelToXMLConverter:
    def __init__(self):
        # Initialize encryption key
        self.encryption_key = os.getenv("ENCRYPTION_KEY", Fernet.generate_key())
        self.cipher_suite = Fernet(self.encryption_key)
        
    def validate_excel(self, file_path: str) -> bool:
        """Validate Excel file format and content."""
        try:
            # Read Excel file
            df = pd.read_excel(file_path)
            
            # Basic validation checks
            if df.empty:
                raise ValueError("Excel file is empty")
                
            # No required columns check; accept any columns
            return True
            
        except Exception as e:
            log.error(f"Excel validation failed: {str(e)}")
            raise ValueError(f"Invalid Excel file: {str(e)}")
    
    def encrypt_data(self, data: str) -> bytes:
        """Encrypt sensitive data using AES-256."""
        try:
            return self.cipher_suite.encrypt(data.encode())
        except Exception as e:
            log.error(f"Encryption failed: {str(e)}")
            raise
    
    def decrypt_data(self, encrypted_data: bytes) -> str:
        """Decrypt encrypted data."""
        try:
            return self.cipher_suite.decrypt(encrypted_data).decode()
        except Exception as e:
            log.error(f"Decryption failed: {str(e)}")
            raise
    
    def sanitize_xml_tag(self, name: str) -> str:
        """Sanitize XML tag name to ensure it's valid."""
        # Replace spaces and special characters with underscores
        tag = re.sub(r'[\s\W]+', '_', name)
        # Remove leading digits or invalid characters
        tag = re.sub(r'^[^a-zA-Z_]', '_', tag)
        # Remove any remaining invalid characters
        tag = re.sub(r'[^a-zA-Z0-9_.-]', '', tag)
        return tag

    def create_xml_header(self, header_fields: Dict[str, Any]) -> ET.Element:
        """Create XML header section from provided fields."""
        header = ET.Element("HEADER")
        
        try:
            for tag_name, tag_value in header_fields.items():
                # Sanitize and validate tag name
                sanitized_tag = self.sanitize_xml_tag(tag_name)
                if not sanitized_tag:
                    raise ValueError(f"Invalid tag name: {tag_name}")
                    
                element = ET.SubElement(header, sanitized_tag)
                element.text = str(tag_value)
                
            return header
            
        except Exception as e:
            log.error(f"Failed to create XML header: {str(e)}")
            raise
    
    def convert_to_xml(self, excel_path: str, header_fields: Dict[str, Any]) -> str:
        """Convert Excel file to XML with custom header."""
        try:
            # Validate Excel file
            self.validate_excel(excel_path)
            
            # Create root element
            root = ET.Element("ROOT")
            
            # Add header
            converter = ExcelToXMLConverter()
            converter._create_header(root, header_fields)
            
            # Read Excel data
            df = pd.read_excel(excel_path)
            
            # Create data section
            data_section = ET.SubElement(root, "DATA")
            
            # Convert each row to XML
            for _, row in df.iterrows():
                record = ET.SubElement(data_section, "RECORD")
                for column in df.columns:
                    # Skip empty values
                    if pd.isna(row[column]):
                        continue
                        
                    field = ET.SubElement(record, column.upper())
                    field.text = str(row[column])
            
            # Pretty print XML
            xml_str = minidom.parseString(ET.tostring(root)).toprettyxml(indent="  ")
            
            # Log success
            log.info("Excel to XML conversion completed successfully")
            
            return xml_str
            
        except Exception as e:
            log.error(f"Conversion failed: {str(e)}")
            raise
    
    def save_xml(self, xml_content: str, output_path: str) -> str:
        """Save XML content to file with encryption."""
        try:
            # Encrypt XML content
            encrypted_content = self.encrypt_data(xml_content)
            
            # Generate unique filename
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"converted_{timestamp}.xml.enc"
            full_path = os.path.join(output_path, filename)
            
            # Save encrypted content
            with open(full_path, "wb") as f:
                f.write(encrypted_content)
            
            log.info(f"XML file saved successfully: {filename}")
            return filename
            
        except Exception as e:
            log.error(f"Failed to save XML file: {str(e)}")
            raise

File: backend/services/converter.js
const XLSX = require('xlsx');
const { create } = require('xmlbuilder2');
const fs = require('fs');
const path = require('path');
const config = require('../config');
const { auditLogger } = require('../utils/logger');

class ExcelToXMLConverter {
    constructor() {
        this.namespace = config.XML_NAMESPACE;
        this.schemaVersion = config.XML_SCHEMA_VERSION;
    }

    /**
     * Sanitize XML tag name to ensure it's valid
     */
    sanitizeXmlTag(tag) {
        if (!tag) {
            return 'EMPTY_TAG';
        }

        // Convert to string and trim
        tag = String(tag).trim();

        // Replace spaces with underscores
        tag = tag.replace(/\s+/g, '_');

        // Remove invalid characters (allow only letters, digits, underscore, hyphen, period)
        tag = tag.replace(/[^a-zA-Z0-9_.-]/g, '');

        // Ensure tag starts with a letter or underscore
        if (!/^[a-zA-Z_]/.test(tag)) {
            tag = '_' + tag;
        }

        // If tag is empty after sanitization, use a fallback
        if (!tag) {
            tag = 'EMPTY_TAG';
        }

        return tag;
    }

    /**
     * Validate Excel file format and content
     */
    validateExcelFile(filePath) {
        try {
            // Check if file exists
            if (!fs.existsSync(filePath)) {
                return false;
            }

            // Check file extension
            const ext = path.extname(filePath).toLowerCase();
            if (!config.ALLOWED_EXTENSIONS.includes(ext)) {
                return false;
            }

            // Try to read the file
            const workbook = XLSX.readFile(filePath);
            
            // Check if there's at least one sheet
            if (!workbook.SheetNames || workbook.SheetNames.length === 0) {
                return false;
            }

            // Get first sheet data
            const sheetName = workbook.SheetNames[0];
            const sheet = workbook.Sheets[sheetName];
            const data = XLSX.utils.sheet_to_json(sheet);

            // Check if file has data
            if (!data || data.length === 0) {
                return false;
            }

            return true;
        } catch (error) {
            auditLogger.logError('system', 'validate_excel_file', error, { file: filePath });
            return false;
        }
    }

    /**
     * Convert Excel file to XML with optional header fields
     */
    convert(inputFile, outputFile, headerFields = {}, sheetName = null, userId = 'system') {
        const startTime = Date.now();

        try {
            // Validate input file exists
            if (!fs.existsSync(inputFile)) {
                throw new Error(`Input file not found: ${inputFile}`);
            }

            // Read Excel file
            const workbook = XLSX.readFile(inputFile);
            
            // Get the appropriate sheet
            const targetSheetName = sheetName || workbook.SheetNames[0];
            if (!workbook.SheetNames.includes(targetSheetName)) {
                throw new Error(`Sheet '${targetSheetName}' not found in Excel file`);
            }

            const sheet = workbook.Sheets[targetSheetName];
            const data = XLSX.utils.sheet_to_json(sheet);

            if (!data || data.length === 0) {
                throw new Error('Excel file is empty');
            }

            // Build XML structure
            const root = create({ version: '1.0', encoding: 'UTF-8' })
                .ele('CALLREPORT');

            // Add header section if header fields are provided
            if (headerFields && Object.keys(headerFields).length > 0) {
                const header = root.ele('HEADER');
                for (const [key, value] of Object.entries(headerFields)) {
                    const tagName = this.sanitizeXmlTag(key);
                    header.ele(tagName).txt(String(value || ''));
                }
            }

            // Create BODY section
            const body = root.ele('BODY');

            // Convert each row to XML
            for (const row of data) {
                const record = body.ele('CALLREPORT_DATA');
                for (const [column, cellValue] of Object.entries(row)) {
                    const tagName = this.sanitizeXmlTag(column);
                    const value = this.formatCellValue(cellValue);
                    record.ele(tagName).txt(value);
                }
            }

            // Convert to pretty-printed XML string
            const xmlContent = root.end({ prettyPrint: true });

            // Write to output file
            fs.writeFileSync(outputFile, xmlContent, 'utf-8');

            // Calculate conversion time
            const conversionTime = Date.now() - startTime;

            // Log successful conversion
            auditLogger.logConversionEvent(
                userId,
                inputFile,
                outputFile,
                conversionTime,
                'success',
                {
                    rows_processed: data.length,
                    columns: Object.keys(data[0] || {}),
                    sheet_name: targetSheetName
                }
            );

            return {
                success: true,
                rowsProcessed: data.length,
                conversionTime
            };

        } catch (error) {
            // Log conversion error
            auditLogger.logError(userId, 'convert_excel_to_xml', error, {
                input_file: inputFile,
                output_file: outputFile,
                sheet_name: sheetName
            });
            throw error;
        }
    }

    /**
     * Format cell value for XML output
     */
    formatCellValue(value) {
        if (value === null || value === undefined) {
            return '';
        }

        if (value instanceof Date) {
            return value.toISOString();
        }

        if (typeof value === 'number') {
            return String(value);
        }

        return String(value);
    }

    /**
     * Get file info from Excel file
     */
    getFileInfo(filePath) {
        try {
            const workbook = XLSX.readFile(filePath);
            const sheetName = workbook.SheetNames[0];
            const sheet = workbook.Sheets[sheetName];
            const data = XLSX.utils.sheet_to_json(sheet);

            return {
                sheetNames: workbook.SheetNames,
                rowCount: data.length,
                columns: Object.keys(data[0] || {})
            };
        } catch (error) {
            throw new Error(`Failed to read Excel file: ${error.message}`);
        }
    }
}

// Create singleton instance
const converter = new ExcelToXMLConverter();

module.exports = { converter, ExcelToXMLConverter };
